### PR TITLE
Returned null in case of empty cookie value

### PR DESCRIPTION
**Pull request by vsin12**
1. some of the old sites by mistake send a blank cookie, and whenever a blank cookie is encountered, failed to parse cookie error comes up - so I made a fix of checking the length and accordingly calling callback with null value.

Note: Either this fix, or CookieJar constructor needs to take SetCookieOptions too, as ignoreError is part of SetCookieOptions interface

### CODE REVIEW

**Comment by ruoho** [comment]
Hi @vsin12, would you be able to update the PR with a test?

**Comment by vsin12** [comment]
@ruoho : Sorry, but I didn't get you ? Do you want me to give a scenario while it was breaking or you want me to write a test to check my pr ?

**Review by awaterma — CHANGES_REQUESTED** [review]
You wrote: `Note: Either this fix, or CookieJar constructor needs to take SetCookieOptions too, as ignoreError is part of SetCookieOptions interface`

ignoreError doesn't apply to storage: `ignoreError - boolean - default false - silently ignore things like parse errors and invalid domains. Store errors aren't ignored by this option` -- so I'm not sure we need to do any more here, in terms of implementation. We might want to start a project to review ignoreError and its applications --- it might be better to have that apply globally, but we'd have to think that through.

**In terms of this P/R; I think I'd like to see the change use validators, as suggested, and then have a simple unit test added that demonstrates the use-case (an empty cookie).**

File: lib/cookie.js [hunk 1]
```diff
@@ -1099,6 +1099,9 @@ class CookieJar {
     if (validators.isFunction(options)) {
       cb = options;
       options = {};
+
+      if(typeof cookie === 'string' && cookie.length <= 0 )
+        return cb(null);
     }
     validators.validate(validators.isFunction(cb), cb);
```
[the remark this answers is not in the document]
Comment (awaterma) [reply lib/cookie.js:1106]: Meaning line 1103/1106 `validators.validate(validators.isFunction(cb), cb);`

**Comment by awaterma** [comment]
I think we still need a unit test for this fix. :)

**Comment by medelibero-sfdc** [comment]
@vsin12 can you change this PR to allow edits from maintainers? I have a few changes I would like to push to this PR.

**Comment by vsin12** [comment]
> @vsin12 can you change this PR to allow edits from maintainers? I have a few changes I would like to push to this PR.

Done.

File: lib/cookie.js [hunk 2]
```diff
@@ -1100,8 +1100,13 @@ class CookieJar {
       cb = options;
       options = {};
     }
+
     validators.validate(validators.isFunction(cb), cb);
 
+    if(!validators.isNonEmptyString(cookie) && !validators.isObject(cookie) && ( cookie instanceof String && cookie.length == 0)) {
+      return cb(null);
+    }
+
```
Comment (awaterma) [inline lib/cookie.js:1109]: Just looked at this in context, and I think this is fine. Will discuss today.

File: test/cookie_jar_test.js [hunk 3]
```diff
@@ -699,4 +699,21 @@ vows
...
+      "with missing parameters": {
+        topic: function() {
+          const jar = new tough.CookieJar();
+          jar.setCookie(
+            '',
+            "https://google.com",
+            this.callback
+          );
+        },
+        "results in a error being returned because of missing parameters": function(err, cookies) {
```
Comment (awaterma) [inline test/cookie_jar_test.js:713]: We don't actually return an error, we are invoking the callback with a Null here. Should we invoke the CB with an Error()?

**Review by awaterma — COMMENTED** [review]
I'll discuss this today; I think we want to add a suggestion to return an error when presented with an empty cookie.

File: lib/cookie.js (hunk 2 above)
Comment (awaterma) [reply lib/cookie.js:1109]: Please see comment below; I think we may want an Error here.

**Review by awaterma — APPROVED** [review]
Approved. We may want to add error checking/throwing at some point, but that's probably in a major release.

**Pull request merged**